lib/Mojo/Server/Hypnotoad.pm

```diff
@@ -0,0 +1,396 @@
+  local $SIG{USR2} = sub { $self->{upgrade} ||= steady_time };
+  $prefork->start;
+  $prefork->daemonize if !$ENV{HYPNOTOAD_FOREGROUND} && $ENV{HYPNOTOAD_REV} < 3;
+
+  # Start accepting connections
+  $prefork->cleanup(1)->run;
+}
+
+sub _cleanup {
+  my ($self, $pid) = @_;
+
+  # Clean up failed upgrade
+  return unless ($self->{new} || '') eq $pid;
+  $self->prefork->app->log->error('Zero downtime software upgrade failed');
+  delete @$self{qw(new upgrade)};
+}
+
+sub _exit { say shift and exit 0 }
+
+sub _finish {
+  my $self = shift;
+
+  $self->{finish} = 1;
+  return unless my $new = $self->{new};
+
+  my $prefork = $self->prefork->cleanup(0);
+  path($prefork->pid_file)->remove;
+  $prefork->ensure_pid_file($new);
+}
+
+sub _hot_deploy {
+
+  # Make sure server is running
+  return unless my $pid = shift->prefork->check_pid;
+
+  # Start hot deployment
+  kill 'USR2', $pid;
+  _exit("Starting hot deployment for Hypnotoad server $pid.");
+}
+
+sub _manage {
+  my $self = shift;
+
+  # Upgraded (wait for all workers to send a heartbeat)
+  my $prefork = $self->prefork;
+  my $log     = $prefork->app->log;
+  if ($ENV{HYPNOTOAD_PID} && $ENV{HYPNOTOAD_PID} ne $$) {
+    return unless $prefork->healthy == $prefork->workers;
+    $log->info("Upgrade successful, stopping $ENV{HYPNOTOAD_PID}");
+    kill 'QUIT', $ENV{HYPNOTOAD_PID};
+  }
+  $ENV{HYPNOTOAD_PID} = $$ unless ($ENV{HYPNOTOAD_PID} // '') eq $$;
+
+  # Upgrade
+  if ($self->{upgrade} && !$self->{finished}) {
+
+    # Fresh start
+    my $ut = $self->upgrade_timeout;
+    unless ($self->{new}) {
+      $log->info("Starting zero downtime software upgrade ($ut seconds)");
+      die "Can't fork: $!" unless defined(my $pid = $self->{new} = fork);
+      exec $^X, $ENV{HYPNOTOAD_EXE} or die "Can't exec: $!" unless $pid;
+    }
+
+    # Timeout
+    kill 'KILL', $self->{new} if $self->{upgrade} + $ut <= steady_time;
+  }
+}
+
+sub _stop {
+  _exit('Hypnotoad server not running.') unless my $pid = shift->prefork->check_pid;
+  kill 'QUIT', $pid;
+  _exit("Stopping Hypnotoad server $pid gracefully.");
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::Server::Hypnotoad - A production web serv...ALL GLORY TO THE HYPNOTOAD!
+
+=head1 SYNOPSIS
+
+  use Mojo::Server::Hypnotoad;
+
+  my $hypnotoad = Mojo::Server::Hypnotoad->new;
+  $hypnotoad->run('/home/sri/myapp.pl');
+
+=head1 DESCRIPTION
+
+L<Mojo::Server::Hypnotoad> is a full featured, UNIX optimized, pre-forking non-blocking I/O HTTP and WebSocket server,
+built around the very well tested and reliable L<Mojo::Server::Prefork>, with IPv6, TLS, SNI, UNIX domain socket, Comet
+(long polling), keep-alive, multiple event loop and hot deployment support that just works. Note that the server uses
+signals for process management, so you should avoid modifying signal handlers in your applications.
+
+To start applications with it you can use the L<hypnotoad> script, which listens on port C<8080>, automatically
```

Comment:
Is iWhy is there not a Mojolicious::Command::hypnotoad module like ::prefork? Is it not possible to be done? If it is possible, what care is needed?